docker/test_docker.py

```diff
@@ -0,0 +1,589 @@
+        """Read and execute permissions are required on /data."""
+        # Note we can't run this one as root because then we have to make a file
+        # that even we can't delete.
+        with tempfile.TemporaryDirectory() as tmpdir:
+            # Make sure the directory is not readable for anyone other than
+            # the owner
+            data_dir = Path(tmpdir) / "data"
+            data_dir.mkdir(mode=perms)
+
+            proc = subprocess.Popen(
+                (
+                    "docker",
+                    "run",
+                    "--rm",
+                    "--user",
+                    # Run as a not us user ID, so access to /data will be
+                    # determined by the "all other users" setting
+                    str(os.getuid() + 1),
+                    "-v",
+                    # Mount the temporary directory as the /data directory
+                    f"{data_dir}:/data",
+                    image,
+                ),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            _, err = proc.communicate()
+
+            # we should error out in this case
+            assert proc.returncode != 0
+            assert "must have read/execute access" in err.decode()
+
+    @pytest.mark.parametrize(
+        "extra_args",
+        (("--user", str(os.getuid())), ("--user", str(os.getuid() + 1))),
+    )
+    def test_needs_rwx_on_packages(self, image: str, extra_args: tuple) -> None:
+        """RWX permission is required on /data/packages."""
+        with tempfile.TemporaryDirectory() as tmpdir:
+            td_path = Path(tmpdir)
+            # Make the /data directory read/writable by anyone
+            td_path.chmod(0o777)
+            # Add the /data/packages directory, and make it readable by anyone,
+            # but writable only by the owner
+            (td_path / "packages").mkdir(mode=0o444)
+
+            proc = subprocess.Popen(
+                (
+                    "docker",
+                    "run",
+                    "--rm",
+                    *extra_args,
+                    "-v",
+                    # Mount the temporary directory as the /data directory
+                    f"{tmpdir}:/data",
+                    image,
+                ),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            _, err = proc.communicate()
+            # we should error out in this case
+            assert proc.returncode != 0
+            assert "must have read/write/execute access" in err.decode()
+
+    def test_runs_as_pypiserver_user(self, image: str) -> None:
+        """Test that the default run uses the pypiserver user."""
+        proc = subprocess.Popen(
+            (
+                "docker",
+                "run",
+                "--rm",
+                "--detach",
+                "--publish",
+                "8080:8080",
```

Comment:
Can we not bind on 8080 on the host but on a random port, someone running the tests may have something else running on that port already